### ISSUE TITLE
Programs

### ISSUE THREAD

**Issue by Rohan200517** [Rohan200517/Statistics#1]
====================================PROGRAM:01=====================================
import pandas as pd

df1 = pd.DataFrame({
    'ID': [1, 2, 3, 4],
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, 35, 40]
})

df2 = pd.DataFrame({
    'ID': [3, 4, 5, 6],
    'Department': ['HR', 'IT', 'Finance', 'Marketing'],
    'Salary': [40000, 50000, 60000, 70000]
})
print(df1)
print(df2)

merged_inner = pd.merge(df1, df2, on='ID', how='inner')
merged_outer = pd.merge(df1, df2, on='ID', how='outer')

print("===== Inner Merge =====")
print(merged_inner)

print("\n===== Outer Merge =====")
print(merged_outer)

df3 = pd.DataFrame({
    'ID': [7, 8],
    'Name': ['Eva', 'Frank'],
    'Age': [28, 33]
})

combined = pd.concat([df1, df3], ignore_index=True)
print("\n===== Concatenated Data =====")
print(combined)

melted = pd.melt(df1,
                 id_vars=['ID'],
                 value_vars=['Name', 'Age'],
                 var_name='Attribute',
                 value_name='Value')

print("\n===== Melted Data =====")
print(melted)

data = pd.DataFrame({
    'Month': ['Jan', 'Jan', 'Feb', 'Feb'],
    'Department': ['HR', 'IT', 'HR', 'IT'],
    'Revenue': [1000, 1500, 1200, 1600]
})

pivoted = data.pivot(index='Month',
                     columns='Department',
                     values='Revenue')

print("\n===== Pivoted Table =====")
print(pivoted)
 
====================================PROGRAM:02=====================================
import re

data = [
    "John Doe, Age:25, Email: john.doe@example.com",
    "Jane Smith, Age:30, Email: jane_smith@company.org",
    "Bob, Age:22, Email: bob22@domain.co.in"
]
print("=== RAW DATA ===")
for d in data:
    print(d)
print("\n=== CLEANED DATA ===")
for record in data:
    record = record.strip().lower()
    name = re.search(r'^[a-z ]+', record)
    age = re.search(r'age[:\s]*(\d+)', record)
    email = re.search(r'[\w\.-]+@[\w\.-]+', record)
    name = name.group().title().strip()
    age = int(age.group(1))
    email = email.group()
    print(f"Name: {name}, Age: {age}, Email: {email}")

====================================PROGRAM:03=====================================
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

data = {
    'Date': pd.date_range(start='2024-01-01', periods=6, freq='M'),
    'Sales': [100, 120, 140, 160, 180, 200],
    'Temperature': [25, 26, 28, 30, 32, 33]
}

df = pd.DataFrame(data)
print("=== Original Time Series Data ===")
print(df)

df['Month'] = df['Date'].dt.month
grouped = df.groupby('Month')['Sales'].sum()
print("\n=== GroupBy: Total Sales per Month ===")
print(grouped)

sales_vector = df['Sales'].values
print("\n=== Data Vector (Sales Only) ===")
print(sales_vector)

multi_vector = df[['Sales', 'Temperature']].values
print("\n=== Multivariate Time Series (Sales & Temperature) ===")
print(multi_vector)

df['Time'] = np.arange(len(df))
X = df[['Time']]
y = df['Sales']

model = LinearRegression()
model.fit(X, y)

future_time = np.arange(len(df), len(df) + 3)
forecast = model.predict(future_time.reshape(-1, 1))

print("\n=== Forecasted Sales for Next 3 Months ===")
print(forecast)
====================================PROGRAM:04=====================================
import numpy as np
from statistics import mode, multimode

print("Enter the number of observations:")
n = int(input())

data = []
freq = []

print("\nEnter data values and their corresponding frequencies:")

for i in range(n):
    x = float(input(f"Data value {i+1}: "))
    f = int(input(f"Frequency of {x}: "))
    data.append(x)
    freq.append(f)

x = np.array(data)
f = np.array(freq)

mean = np.sum(x * f) / np.sum(f)

cf = np.cumsum(f)
N = np.sum(f)
median_pos = (N + 1) / 2

for i in range(len(cf)):
    if cf[i] >= median_pos:
        median = x[i]
        break

modal_value = x[np.argmax(f)]

range_value = np.max(x) - np.min(x)

mean_diff_sq = np.sum(f * (x - mean)**2)
variance = mean_diff_sq / np.sum(f)
std_dev = np.sqrt(variance)

mean_deviation = np.sum(f * np.abs(x - mean)) / np.sum(f)

Q1_pos = N / 4
Q3_pos = 3 * N / 4

def find_quartile_value(cf, x, qpos):
    for i in range(len(cf)):
        if cf[i] >= qpos:
            return x[i]

Q1 = find_quartile_value(cf, x, Q1_pos)
Q3 = find_quartile_value(cf, x, Q3_pos)
quartile_deviation = (Q3 - Q1) / 2

print("\n===== RESULTS =====")
print(f"Mean = {mean:.2f}")
print(f"Median = {median:.2f}")
print(f"Mode = {modal_value:.2f}")
print(f"Range = {range_value:.2f}")
print(f"Variance = {variance:.2f}")
print(f"Standard Deviation = {std_dev:.2f}")
print(f"Mean Deviation = {mean_deviation:.2f}")
print(f"Quartile Deviation = {quartile_deviation:.2f}")

====================================PROGRAM:05=====================================
from sklearn.model_selection import train_test_split, KFold, LeaveOneOut, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.datasets import fetch_california_housing
import numpy as np
import warnings
warnings.filterwarnings("ignore", message="R^2 score is not well-defined*")

X, y = fetch_california_housing(return_X_y=True)
model = LinearRegression()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.8, random_state=0)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print("Validation Set:")
print("RMSE =", np.sqrt(mean_squared_error(y_test, y_pred)))
print("MAE  =", mean_absolute_error(y_test, y_pred))
print("R2   =", r2_score(y_test, y_pred))

loo = LeaveOneOut()
rmse = np.sqrt(-cross_val_score(model, X, y, cv=loo, scoring='neg_mean_squared_error').mean())
mae = -cross_val_score(model, X, y, cv=loo, scoring='neg_mean_absolute_error').mean()
r2  = cross_val_score(model, X, y, cv=loo, scoring='r2').mean()
print("\nLOOCV:")
print("RMSE =", rmse)
print("MAE  =", mae)
print("R2   =", r2)

kfold = KFold(n_splits=5, shuffle=True, random_state=0)
rmse = np.sqrt(-cross_val_score(model, X, y, cv=kfold, scoring='neg_mean_squared_error').mean())
mae = -cross_val_score(model, X, y, cv=kfold, scoring='neg_mean_absolute_error').mean()
r2  = cross_val_score(model, X, y, cv=kfold, scoring='r2').mean()
print("\nK-Fold Cross-Validation:")
print("RMSE =", rmse)
print("MAE  =", mae)
print("R2   =", r2)

====================================PROGRAM:06=====================================
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm,binom,poisson,bernoulli

data=[2,3,3,4,4,5,5,6,6,7]

mean = np.mean(data)
var = np.var(data)
n = len(data)
p = 0.5
lam = mean

print("Mean=",mean)
print("Variance=",var)

x=np.linspace(min(data),max(data),100)
plt.subplot(2,2,1)
plt.plot(x,norm.pdf(x,mean,np.sqrt(var)),color='r')
plt.title("Normal Distribution")

x=np.arange(0,n+1)
plt.subplot(2,2,2)
plt.bar(x,binom.pmf(x,n,p),color='b')
plt.title("Binomial Distribution")

x=np.arange(0,max(data)+3)
plt.subplot(2,2,3)
plt.bar(x,poisson.pmf(x,lam),color='g')
plt.title("Poisson Distribution")

x=[0,1]
plt.subplot(2,2,4)
plt.bar(x,bernoulli.pmf(x,p),color='orange')
plt.title("Bernoulli Distribution")

plt.tight_layout()
plt.show()

print("\nNormal:Continuous bell-shaped curve:")
print("\nBinomial:Fixed trails with success probability:")
print("\nPoisson:Counts of rare events:")
print("\nBernoulli:Single trail success/failure:")

====================================PROGRAM:07=====================================

import numpy as np
from scipy import stats

sample1 = np.array([12, 15, 14, 10, 13, 12, 14])
sample2 = np.array([10, 9, 12, 8, 9, 11, 10])
paired_before = np.array([20, 22, 19, 24, 18])
paired_after  = np.array([21, 20, 20, 23, 19])

mu0 = 12
t1, p1 = stats.ttest_1samp(sample1, mu0)
print("=== One-Sample t-Test ===")
print(f"t = {t1:.3f}, p = {p1:.4f}")
print("=> " + ("Significant difference from mean" if p1 < 0.05 else "No significant difference"))

t2, p2 = stats.ttest_ind(sample1, sample2, equal_var=False)
print("\n=== Two-Sample t-Test ===")
print(f"t = {t2:.3f}, p = {p2:.4f}")
print("=> " + ("Samples differ significantly" if p2 < 0.05 else "No significant difference between samples"))

t3, p3 = stats.ttest_rel(paired_before, paired_after)
print("\n=== Paired-Sample t-Test ===")
print(f"t = {t3:.3f}, p = {p3:.4f}")
print("=> " + ("Significant change (before vs after)" if p3 < 0.05 else "No significant change")) 

====================================PROGRAM:08=====================================

import pandas as pd
from scipy import stats
import statsmodels.api as sm
from statsmodels.formula.api import ols

group1 = [85, 89, 86, 88, 90]
group2 = [78, 74, 77, 72, 75]
group3 = [82, 85, 84, 83, 86]

f, p = stats.f_oneway(group1, group2, group3)

print("----- ONE-WAY ANOVA -----")
print("F-value:", round(f, 3))
print("P-value:", round(p, 3))

if p < 0.05:
    print("Result: Significant difference between groups\n")
else:
    print("Result: No significant difference between groups\n")

data = {
    'Score': [85, 89, 86, 88, 90, 78, 74, 77, 72, 75,
              82, 85, 84, 83, 86, 88, 90, 91, 87, 89],
    'Method': ['A'] * 10 + ['B'] * 10,
    'Gender': ['Male', 'Female'] * 10
}

df = pd.DataFrame(data)

model = ols('Score ~ C(Method) + C(Gender) + C(Method):C(Gender)', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)

print("----- TWO-WAY ANOVA -----")
print(anova_table)

====================================PROGRAM:09=====================================

import numpy as np
import pandas as pd
from scipy.stats import pearsonr, spearmanr
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
import seaborn as sns

x = np.array([10, 20, 30, 40, 50, 60])
y = np.array([15, 25, 35, 45, 55, 65])

df = pd.DataFrame({"X": x, "Y": y})

pear_corr, _ = pearsonr(df["X"], df["Y"])
print("Pearson Correlation:", pear_corr)

rank_corr, _ = spearmanr(df["X"], df["Y"])
print("Spearman Rank Correlation:", rank_corr)

X_reshaped = df[["X"]]
model = LinearRegression()
model.fit(X_reshaped, df["Y"])

print("\nRegression Equation: Y = {:.2f}X + {:.2f}".format(model.coef_[0], model.intercept_))

plt.figure(figsize=(6, 4))
plt.scatter(df["X"], df["Y"], label="Data Points")
plt.plot(df["X"], model.predict(X_reshaped), label="Regression Line")
plt.xlabel("X")
plt.ylabel("Y")
plt.title("X-Y Scatter Plot with Regression Line")
plt.legend()
plt.grid(True)
plt.show()

plt.figure(figsize=(5, 4))
sns.heatmap(df.corr(), annot=True, cmap="coolwarm")
plt.title("Correlation Matrix Heatmap")
plt.show()

====================================PROGRAM:10=====================================

import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
import seaborn as sns

data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target)

print("Shape of Dataset:", X.shape)

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)

print("\nExplained Variance Ratio:")
print(pca.explained_variance_ratio_)

pca_df = pd.DataFrame(X_pca, columns=["PC1", "PC2"])
pca_df["Class"] = y.map({0: "Malignant", 1: "Benign"})

plt.figure(figsize=(7, 5))
sns.scatterplot(
    x="PC1", y="PC2", hue="Class",
    palette=["red", "green"], data=pca_df, alpha=0.7
)
plt.title("PCA: Wisconsin Breast Cancer Dataset")
plt.xlabel("Principal Component 1")
plt.ylabel("Principal Component 2")
plt.grid(True)
plt.show()

pca_full = PCA()
pca_full.fit(X_scaled)

plt.figure(figsize=(7, 5))
plt.plot(np.cumsum(pca_full.explained_variance_ratio_), marker="o")
plt.title("Cumulative Explained Variance by PCA Components")
plt.xlabel("Number of Components")
plt.ylabel("Cumulative Explained Variance")
plt.grid(True)
plt.show()

====================================PROGRAM:11=====================================

import numpy as np
import matplotlib.pyplot as plt
from sklearn import datasets
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import StandardScaler

iris = datasets.load_iris()
X = iris.data
y = iris.target
target_names = iris.target_names

print("Iris dataset shape:", X.shape)

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

lda = LinearDiscriminantAnalysis(n_components=2)
X_lda = lda.fit_transform(X_scaled, y)

print("Transformed LDA shape:", X_lda.shape)

plt.figure(figsize=(10, 6))
colors = ["red", "green", "blue"]

for color, i, target in zip(colors, [0, 1, 2], target_names):
    plt.scatter(
        X_lda[y == i, 0],
        X_lda[y == i, 1],
        alpha=0.7,
        color=color,
        label=target
    )

plt.xlabel("LDA Component 1")
plt.ylabel("LDA Component 2")
plt.title("Linear Discriminant Analysis on Iris Dataset")
plt.legend()
plt.grid(True)
plt.show()

print("\nExplained variance ratio:")
print(lda.explained_variance_ratio_)

====================================PROGRAM:12=====================================

import numpy as np
import matplotlib.pyplot as plt
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

iris = datasets.load_iris()
X = iris.data[:, [0, 1, 3]]
y = iris.data[:, 2]

print("Feature shape:", X.shape)
print("Target shape:", y.shape)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print("\n=========== Regression Results ===========")
print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)
print("Mean Squared Error:", mean_squared_error(y_test, y_pred))
print("R² Score:", r2_score(y_test, y_pred))

plt.figure(figsize=(10, 5))
plt.scatter(y_test, y_pred)
plt.plot(
    [min(y_test), max(y_test)],
    [min(y_test), max(y_test)],
    linewidth=2
)
plt.xlabel("Actual Petal Length")
plt.ylabel("Predicted Petal Length")
plt.title("Actual vs Predicted Petal Length (Multiple Linear Regression)")
plt.grid(True)
plt.show()

residuals = y_test - y_pred

plt.figure(figsize=(10, 5))
plt.scatter(y_pred, residuals)
plt.axhline(0, linewidth=2)
plt.xlabel("Predicted Petal Length")
plt.ylabel("Residuals")
plt.title("Residual Plot")
plt.grid(True)
plt.show()
